specs/WebResourceRequested-CustomScheme.md

```diff
@@ -0,0 +1,295 @@
+// The WebResourceRequested event will not be raised for this request.
+webView.CoreWebView2.ExecuteScriptAsync("var oReq = new XMLHttpRequest();\
+                                        oReq.addEventListener(\"load\", reqListener);\
+                                        oReq.open(\"GET\", \"custom-scheme-not-in-allowed-origins://example-data.json\");\
+                                        oReq.send();");
+```
+
+``` cpp
+Microsoft::WRL::ComPtr<ICoreWebView2EnvironmentOptions3> options3;
+if (options.As(&options3) == S_OK) {
+  std::vector<Microsoft::WRL::ComPtr<ICoreWebView2CustomSchemeRegistration>> schemeRegistrations;
+  
+  const WCHAR* allowedOrigins[1] = {L"https://*.example.com"};
+  schemeRegistrations.push_back(Microsoft::WRL::Make<CoreWebView2CustomSchemeRegistration>(
+          L"custom-scheme", TRUE /* isSecure*/, 1, allowedOrigins));
+  schemeRegistrations.push_back(Microsoft::WRL::Make<CoreWebView2CustomSchemeRegistration>(
+          L"custom-scheme-not-in-allowed-origins-list", TRUE /* isSecure*/, nullptr));
+  CHECK_FAILURE(options3->SetCustomSchemeRegistrations(schemeRegistrations.size(), schemeRegistrations.data());
+}
+...
+
+CHECK_FAILURE(m_webView->AddWebResourceRequestedFilter(
+        L"custom-scheme*", COREWEBVIEW2_WEB_RESOURCE_CONTEXT_ALL));
+CHECK_FAILURE(m_webView->add_WebResourceRequested(
+    Callback<ICoreWebView2WebResourceRequestedEventHandler>(
+        [this](
+            ICoreWebView2* sender,
+            ICoreWebView2WebResourceRequestedEventArgs* args) {
+        wil::com_ptr<ICoreWebView2WebResourceRequest> request;
+        wil::com_ptr<ICoreWebView2WebResourceResponse> response;
+        CHECK_FAILURE(args->get_Request(&request));
+        wil::unique_cotaskmem_string uri;
+        CHECK_FAILURE(request->get_Uri(&uri));
+        if (wcsncmp(uri.get(), L"custom-scheme", ARRAYSIZE(L"custom-scheme")-1) == 0)
+        {
+            std::wstring assetsFilePath = L"data/";
+            assetsFilePath += wcsstr(uri.get(), L"://") + 3;
+            wil::com_ptr<IStream> stream;
+            SHCreateStreamOnFileEx(
+                assetsFilePath.c_str(), STGM_READ, FILE_ATTRIBUTE_NORMAL, FALSE,
+                nullptr,
+                &stream);
+            if (stream)
+            {
+                CHECK_FAILURE(
+                    m_appWindow->GetWebViewEnvironment()->CreateWebResourceResponse(
+                        stream.get(), 200, L"OK", L"Content-Type: application/json\nAccess-Control-Allow-Origin: *",
+                        &response));
+                CHECK_FAILURE(args->put_Response(response.get()));
+            }
+            else
+            {
+                CHECK_FAILURE(
+                    m_appWindow->GetWebViewEnvironment()->CreateWebResourceResponse(
+                        nullptr, 404, L"Not Found", L"", &response));
+                CHECK_FAILURE(args->put_Response(response.get()));
+            }
+            return S_OK;
+        }
+
+        return S_OK;
```

Comment:
I assume we're returning `S_OK` so that another event handler has a chance to respond.

Reply:
Return value is ignored. Response property being set or not is how event handlers can decide who is providing the response.